Fix data race on shutdown

zmb3 opened the pull request:
Updates #14143

zmb3 committed:
Fix data race on shutdown

Updates #14143

espadolini commented on lib/service/service.go:
```diff
@@ -2339,52 +2358,40 @@ func (process *TeleportProcess) initSSH() error {
+			waitFn = agentPool.Wait
+
+			process.OnExit("agentpool.shutdown", func(payload interface{}) {
+				agentPool.Stop()
```
Stopping the agentPool will kill all reverse tunnel connections, so this should happen after the server has been waited, while this will happen at the same time as the other shutdowns.

Considering that `*AgentPool`'s `Stop` already internally handles having a nil pointer (by doing nothing) maybe we can just pass the optional `agentPool` pointer into `ssh.shutdown` like it was done before? Moreover, we probably want to handle hard shutdown vs graceful shutdown vs graceful restart differently in the future, and that info is probably only going to be available in the `ssh.shutdown` callback anyway.

espadolini commented on lib/service/service.go:
```diff
@@ -2103,22 +2103,21 @@ func (process *TeleportProcess) initSSH() error {
 			return trace.Wrap(err)
 		}
 
+		ok := false
```
What do you think of using a named `err` return value and checking it for `nil`? Too easy to screw up?

zmb3 replied:
I think so. Easy to accidentally shadow.

espadolini commented on lib/service/service.go:
```diff
@@ -2339,52 +2358,40 @@ func (process *TeleportProcess) initSSH() error {
-			log.Infof("Shutting down immediately.")
-			if s != nil {
+		// Execute this when process is asked to exit.
+		process.OnExit("ssh.shutdown", func(payload interface{}) {
```
It's tradition to do it like this, but there's nothing inherently special about `OnExit`. We could just block waiting for `process.WaitForEvent(TeleportExitEvent)` inline and do the cleanup afterwards right in the `ssh.node` callback, leaving some (or most?) of the defers to do the work. WDYT?

zmb3 replied:
That might be a bit simpler. Will give this a shot.

espadolini commented on lib/service/service.go:
```diff
@@ -4219,14 +4229,20 @@ func (process *TeleportProcess) initApps() {
 
+		defer func() {
+			if !ok {
+				warnOnErr(appServer.Close(), log)
```
Considering that the whole point of `warnOnErr` is to avoid warning on double close, perhaps we can just `defer func() { warnOnErr(conn.Close()) }` and `defer func() { warnOnErr(appServer.Close()) }` blindly, and not even bother with the `conn.Close` in the `app.shutdown` func?

zmb3 replied:
Just to be clear, are you suggesting the same thing as your other comment? (removing the `OnExit` and just deferring the cleanup?)

If not, I'm not sure I understand this suggestion. The point of the deferred close here is only to clean up in the error case (when we don't register the onexit handler).

espadolini replied:
I'm saying that we can just defer the close regardless of whether or not we actually succeed in initializing the app service (and thus close during the shutdown); if you prefer to keep the `ok` flag then that's fine too.

zmb3 replied:
But the app service's close is on a separate goroutine, so we'd have to fix that first, otherwise we will prematurely close things, right?

espadolini replied:
Well, `apps.start` will wait until server and agent pool are both closed anyway - it's all very coincidental tho, so I agree that it's better to be explicit with the flag (or with a wait for `TeleportExitEvent`).

zmb3 committed:
Restore agentpool.Stop()

Both Stop and Wait methods handle a nil receiver, so this is safe
and ensures that we stop the agent pool only after the SSH server
shuts down.

zmb3 committed:
Remove the ssh.shutdown handler

Instead, just block until the server terminates and clean up
in line.

zmb3 commented on lib/service/service.go:
```diff
@@ -2273,6 +2270,7 @@ func (process *TeleportProcess) initSSH() error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
+		defer func() { warnOnErr(s.Close(), log) }()
```
@espadolini does this look safe?

If there's an error and we return early, this close will do an immediate shutdown.

If there's no error and everything starts up successfully, then the clean up at the end of this function will run, and my hope is that the deferred `Close` call will effectively be a no op.

espadolini replied:
I think a double close on most of our networking stuff is either a noop or is harmless but returns a `net.ErrClosed` or equivalent (and thus gets ignored by `warnOnErr`) so we're good here.

zmb3 commented on lib/service/service.go:
```diff
@@ -2160,18 +2154,20 @@ func (process *TeleportProcess) initSSH() error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
+		// TODO: are we missing rm.Close()
```
This, as well as `authClient` and `storagePresence` are never closed. Any idea if this was intentional or an oversight?

espadolini replied:
`storagePresence` is a wrapper around process storage, which is closed after all services are done during shutdown, so we must not close that.

`authClient` will close itself after `process.ExitContext()` is done; we don't explicitly close any other cache as returned by `process.newAccessCache` and I'm not sure if we should start.

zmb3 replied:
Ok, what about `rm` - should it be closed?

cc @jakule - I think the restricted session manager is leaky either way. Could you take a look? (even if we do call `rm.Close()`, nothing appears to close `l.events.EventCh`, so the `auditEventLoop` never terminates.

espadolini commented on lib/service/service.go:
```diff
@@ -2339,52 +2335,31 @@ func (process *TeleportProcess) initSSH() error {
-		}
+		agentPool.Wait()
 
-		log.Infof("Exited.")
-		return nil
-	})
+		event, err := process.WaitForEvent(process.ExitContext(), TeleportExitEvent)
```
The `s.Wait()` and `agentPool.Wait()` calls should go after the `WaitForEvent` and the close/shutdown, even though technically `s.Wait()` can return on a network error on the listener, but then we'd still get stuck on `agentPool.Wait()` anyway, which will never exit until the agent pool is explicitly stopped.

zmb3 committed:
Move server wait after exit event is recieved

zmb3 commented:
I think we've addressed everything. Any further comments, @espadolini?

espadolini commented on lib/service/service.go:
```diff
@@ -2339,52 +2335,31 @@ func (process *TeleportProcess) initSSH() error {
-			warnOnErr(conn.Close(), log)
-		}
+		s.Wait()
+		agentPool.Wait()
+		agentPool.Stop()
```
```suggestion
		agentPool.Stop()
		agentPool.Wait()
```

😅

zmb3 replied:
Phew. What would I do without you.

rosstimothy reviewed (approved):
LGTM barring the last suggestion

zmb3 committed:
Stop and Wait

espadolini committed:
Merge branch 'master' into zmb3/fix-shutdown-race

espadolini reviewed (approved):
Fatfingered "update branch", sorry

zmb3 committed:
Merge branch 'master' into zmb3/fix-shutdown-race
